pkg/bmh/list.go

```diff
@@ -0,0 +1,107 @@
+package bmh
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	bmhv1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+)
+
+const (
+	fiveScds time.Duration = 5 * time.Second
+)
+
+// List returns bareMetalHosts inventory in the given namespace.
+func List(apiClient *clients.Settings, nsname string) ([]*BmhBuilder, error) {
+	glog.V(100).Infof("Listing bareMetalHosts in the nsname %s", nsname)
+
+	if nsname == "" {
+		glog.V(100).Infof("bareMetalHost 'nsname' parameter can not be empty")
+
+		return nil, fmt.Errorf("failed to list bareMetalHosts, 'nsname' parameter is empty")
+	}
+
+	var bmhList bmhv1alpha1.BareMetalHostList
+	err := apiClient.List(context.TODO(), &bmhList, &goclient.ListOptions{Namespace: nsname})
+
+	if err != nil {
+		glog.V(100).Infof("Failed to list bareMetalHosts in the nsname %s due to %s", nsname, err.Error())
+
+		return nil, err
+	}
+
+	if len(bmhList.Items) < 1 {
+		glog.V(100).Infof("Cluster doesn't have baremetalhosts deployed, bad cluster state ...")
+
+		return nil, fmt.Errorf("BaremetaHosts list is empty")
+	}
+
+	var bmhObjects []*BmhBuilder
+
+	for _, baremetalhost := range bmhList.Items {
+		copiedBmh := baremetalhost
+		bmhBuilder := &BmhBuilder{
+			apiClient:  apiClient,
+			Object:     &copiedBmh,
+			Definition: &copiedBmh,
+		}
+
+		bmhObjects = append(bmhObjects, bmhBuilder)
+	}
+
+	return bmhObjects, nil
+}
+
+// WaitForAllBareMetalHostsInGoodOperationalState waits for all baremetalhosts to be in good Operational State
+// for a time duration up to the timeout.
+func WaitForAllBareMetalHostsInGoodOperationalState(apiClient *clients.Settings,
```

Comment:
Please add an goclient.ListOptions argument so that users can be more granular regarding the BMHs that they want to list

Reply:
not needed, I don't see it could be useful. and, anyway, I add only methods and options I could test. If anyone will need more and will have case an environment to test it so he could add such option.

Reply:
It's useful because some folks may need to only get a subset of BMHs based on things such as LabelSelectors or FieldSelectors. Typically I agree that you should only add what you need but in this case changing it later what cause you to have to refactor any tests that use it. Please reference this [function](https://github.com/openshift-kni/eco-goinfra/blob/main/pkg/pod/list.go#L76) as an example.

Reply:
Maybe I was not clear; let me explain it again.
There is no use case for the bmh with options. After 4 years with ocp-edge-auto, I can tell you we don't need it.
In addition, it will be a bit complicated to implement; on the primary List method, we are already using a goclient.ListOptions to filter bmh list by namespace (at least for me on golang, I don't understand how to do so)
The primary usage for a bmh object is to work directly with a specific bmh object, a single one. The list is mainly used to overview all existing bmh, useful in case of scale up/down, etcd, replacement to choose random bmh. You could look at the bmh object to understand that there is nothing to filter out except the namespace.